cpp/include/raft/linalg/detail/transpose.cuh

```diff
@@ -49,7 +51,7 @@ RAFT_KERNEL transpose_half_kernel(IndexType n_rows,
 
       for (int j = 0; j < TILE_DIM; j += BLOCK_ROWS) {
         if (x < n_cols && (y + j) < n_rows) {
-          tile[threadIdx.y + j][threadIdx.x] = __ldg(&in[(y + j) * n_cols + x]);
+          tile[threadIdx.y + j][threadIdx.x] = __ldg(&in[(y + j) * stride_in + x]);
```

Comment:
It's advisable to use [raft's helpers](https://github.com/rapidsai/raft/blob/ac53a0fdc35fe36a11a3ac4debd4cc9a4076fe7f/cpp/include/raft/util/device_loads_stores.cuh#L495) for this unless you need a specific cache behavior (for which, maybe, we should add more helpers?..)
```suggestion
          tile[threadIdx.y + j][threadIdx.x] = raft::ldg(&in[(y + j) * stride_in + x]);
```

Reply:
Hi @achirkin , thank you for your suggestion! Yeah, you're right! Because `cublas<t>geam()`does not support the half. About the `cublasLtMatrixTransform`. Let me see how many I can change. (would you like to suggest changing this PR or the next separate one to change all of `transpose` ? )

Reply:
Hi, Dear @achirkin; I just tested the `cublasLtMatrixTransform` benchmark and found that performance would be lower than the current implementation by around 10%. So, could I keep the current implementation temporarily:

- Current `transpose_half_kernel`
```
TransposeBench<half, int, raft::row_major>/0/manual_time        0.006 ms        0.041 ms       125637 10#10000
TransposeBench<half, int, raft::row_major>/1/manual_time        0.016 ms        0.052 ms        42913 10#100000
TransposeBench<half, int, raft::row_major>/2/manual_time        0.122 ms        0.158 ms         5774 10#1000000
TransposeBench<half, int, raft::row_major>/3/manual_time        0.011 ms        0.046 ms        65254 128#10000
TransposeBench<half, int, raft::row_major>/4/manual_time        0.084 ms        0.120 ms         8361 128#100000
TransposeBench<half, int, raft::row_major>/5/manual_time        0.762 ms        0.798 ms          917 128#1000000
TransposeBench<half, int, raft::row_major>/6/manual_time        0.022 ms        0.058 ms        31617 256#10000
TransposeBench<half, int, raft::row_major>/7/manual_time        0.156 ms        0.192 ms         4486 256#100000
TransposeBench<half, int, raft::row_major>/8/manual_time         1.53 ms         1.57 ms          454 256#1000000
TransposeBench<half, int, raft::row_major>/9/manual_time        0.035 ms        0.071 ms        20037 512#10000
TransposeBench<half, int, raft::row_major>/10/manual_time       0.310 ms        0.346 ms         2264 512#100000
TransposeBench<half, int, raft::row_major>/11/manual_time        3.09 ms         3.14 ms          225 512#1000000
TransposeBench<half, int, raft::row_major>/12/manual_time       0.073 ms        0.108 ms         9565 1024#10000
TransposeBench<half, int, raft::row_major>/13/manual_time       0.642 ms        0.678 ms         1088 1024#100000
TransposeBench<half, int, raft::row_major>/14/manual_time        6.29 ms         6.36 ms          105 1024#1000000

TransposeBench<half, int, raft::col_major>/0/manual_time        0.006 ms        0.041 ms       124832 10#10000
TransposeBench<half, int, raft::col_major>/1/manual_time        0.017 ms        0.052 ms        42333 10#100000
TransposeBench<half, int, raft::col_major>/2/manual_time        0.125 ms        0.161 ms         5659 10#1000000
TransposeBench<half, int, raft::col_major>/3/manual_time        0.011 ms        0.046 ms        65143 128#10000
TransposeBench<half, int, raft::col_major>/4/manual_time        0.084 ms        0.120 ms         8351 128#100000
TransposeBench<half, int, raft::col_major>/5/manual_time        0.762 ms        0.798 ms          916 128#1000000
TransposeBench<half, int, raft::col_major>/6/manual_time        0.022 ms        0.058 ms        31497 256#10000
TransposeBench<half, int, raft::col_major>/7/manual_time        0.156 ms        0.192 ms         4484 256#100000
TransposeBench<half, int, raft::col_major>/8/manual_time         1.53 ms         1.57 ms          454 256#1000000
TransposeBench<half, int, raft::col_major>/9/manual_time        0.035 ms        0.071 ms        19983 512#10000
TransposeBench<half, int, raft::col_major>/10/manual_time       0.310 ms        0.346 ms         2263 512#100000
TransposeBench<half, int, raft::col_major>/11/manual_time        3.09 ms         3.14 ms          225 512#1000000
TransposeBench<half, int, raft::col_major>/12/manual_time       0.073 ms        0.109 ms         9534 1024#10000
TransposeBench<half, int, raft::col_major>/13/manual_time       0.643 ms        0.679 ms         1087 1024#100000
TransposeBench<half, int, raft::col_major>/14/manual_time        6.29 ms         6.36 ms          105 1024#1000000
```
- Current `transpose_half_kernel`
```
TransposeBench<half, int, raft::row_major>/0/manual_time        0.006 ms        0.041 ms       125637 10#10000
TransposeBench<half, int, raft::row_major>/1/manual_time        0.016 ms        0.052 ms        42913 10#100000
TransposeBench<half, int, raft::row_major>/2/manual_time        0.122 ms        0.158 ms         5774 10#1000000
TransposeBench<half, int, raft::row_major>/3/manual_time        0.011 ms        0.046 ms        65254 128#10000
TransposeBench<half, int, raft::row_major>/4/manual_time        0.084 ms        0.120 ms         8361 128#100000
TransposeBench<half, int, raft::row_major>/5/manual_time        0.762 ms        0.798 ms          917 128#1000000
TransposeBench<half, int, raft::row_major>/6/manual_time        0.022 ms        0.058 ms        31617 256#10000
TransposeBench<half, int, raft::row_major>/7/manual_time        0.156 ms        0.192 ms         4486 256#100000
TransposeBench<half, int, raft::row_major>/8/manual_time         1.53 ms         1.57 ms          454 256#1000000
TransposeBench<half, int, raft::row_major>/9/manual_time        0.035 ms        0.071 ms        20037 512#10000
TransposeBench<half, int, raft::row_major>/10/manual_time       0.310 ms        0.346 ms         2264 512#100000
TransposeBench<half, int, raft::row_major>/11/manual_time        3.09 ms         3.14 ms          225 512#1000000
TransposeBench<half, int, raft::row_major>/12/manual_time       0.073 ms        0.108 ms         9565 1024#10000
TransposeBench<half, int, raft::row_major>/13/manual_time       0.642 ms        0.678 ms         1088 1024#100000
TransposeBench<half, int, raft::row_major>/14/manual_time        6.29 ms         6.36 ms          105 1024#1000000

TransposeBench<half, int, raft::col_major>/0/manual_time        0.006 ms        0.041 ms       124832 10#10000
TransposeBench<half, int, raft::col_major>/1/manual_time        0.017 ms        0.052 ms        42333 10#100000
TransposeBench<half, int, raft::col_major>/2/manual_time        0.125 ms        0.161 ms         5659 10#1000000
TransposeBench<half, int, raft::col_major>/3/manual_time        0.011 ms        0.046 ms        65143 128#10000
TransposeBench<half, int, raft::col_major>/4/manual_time        0.084 ms        0.120 ms         8351 128#100000
TransposeBench<half, int, raft::col_major>/5/manual_time        0.762 ms        0.798 ms          916 128#1000000
TransposeBench<half, int, raft::col_major>/6/manual_time        0.022 ms        0.058 ms        31497 256#10000
TransposeBench<half, int, raft::col_major>/7/manual_time        0.156 ms        0.192 ms         4484 256#100000
TransposeBench<half, int, raft::col_major>/8/manual_time         1.53 ms         1.57 ms          454 256#1000000
TransposeBench<half, int, raft::col_major>/9/manual_time        0.035 ms        0.071 ms        19983 512#10000
TransposeBench<half, int, raft::col_major>/10/manual_time       0.310 ms        0.346 ms         2263 512#100000
TransposeBench<half, int, raft::col_major>/11/manual_time        3.09 ms         3.14 ms          225 512#1000000
TransposeBench<half, int, raft::col_major>/12/manual_time       0.073 ms        0.109 ms         9534 1024#10000
TransposeBench<half, int, raft::col_major>/13/manual_time       0.643 ms        0.679 ms         1087 1024#100000
TransposeBench<half, int, raft::col_major>/14/manual_time        6.29 ms         6.36 ms          105 1024#1000000
```
- `cublasLtMatrixTransform` version
```
TransposeBench<half, int, raft::row_major>/0/manual_time        0.009 ms        0.044 ms        82849 10#10000
TransposeBench<half, int, raft::row_major>/1/manual_time        0.017 ms        0.053 ms        40498 10#100000
TransposeBench<half, int, raft::row_major>/2/manual_time        0.108 ms        0.144 ms         6468 10#1000000
TransposeBench<half, int, raft::row_major>/3/manual_time        0.014 ms        0.049 ms        49384 128#10000
TransposeBench<half, int, raft::row_major>/4/manual_time        0.091 ms        0.127 ms         7724 128#100000
TransposeBench<half, int, raft::row_major>/5/manual_time        0.845 ms        0.882 ms          828 128#1000000
TransposeBench<half, int, raft::row_major>/6/manual_time        0.023 ms        0.059 ms        30710 256#10000
TransposeBench<half, int, raft::row_major>/7/manual_time        0.186 ms        0.222 ms         3775 256#100000
TransposeBench<half, int, raft::row_major>/8/manual_time         1.80 ms         1.84 ms          389 256#1000000
TransposeBench<half, int, raft::row_major>/9/manual_time        0.041 ms        0.076 ms        17249 512#10000
TransposeBench<half, int, raft::row_major>/10/manual_time       0.395 ms        0.431 ms         1772 512#100000
TransposeBench<half, int, raft::row_major>/11/manual_time        3.91 ms         3.95 ms          179 512#1000000
TransposeBench<half, int, raft::row_major>/12/manual_time       0.076 ms        0.112 ms         9123 1024#10000

TransposeBench<half, int, raft::row_major>/13/manual_time       0.796 ms        0.832 ms          878 1024#100000
TransposeBench<half, int, raft::row_major>/14/manual_time        7.94 ms         8.03 ms           84 1024#1000000
TransposeBench<half, int, raft::col_major>/0/manual_time        0.009 ms        0.044 ms        82791 10#10000
TransposeBench<half, int, raft::col_major>/1/manual_time        0.017 ms        0.053 ms        40396 10#100000
TransposeBench<half, int, raft::col_major>/2/manual_time        0.109 ms        0.145 ms         6415 10#1000000
TransposeBench<half, int, raft::col_major>/3/manual_time        0.014 ms        0.049 ms        49297 128#10000
TransposeBench<half, int, raft::col_major>/4/manual_time        0.091 ms        0.127 ms         7714 128#100000
TransposeBench<half, int, raft::col_major>/5/manual_time        0.847 ms        0.884 ms          826 128#1000000
TransposeBench<half, int, raft::col_major>/6/manual_time        0.023 ms        0.059 ms        30682 256#10000
TransposeBench<half, int, raft::col_major>/7/manual_time        0.186 ms        0.222 ms         3771 256#100000
TransposeBench<half, int, raft::col_major>/8/manual_time         1.80 ms         1.84 ms          388 256#1000000
TransposeBench<half, int, raft::col_major>/9/manual_time        0.041 ms        0.077 ms        17233 512#10000
TransposeBench<half, int, raft::col_major>/10/manual_time       0.396 ms        0.432 ms         1770 512#100000
TransposeBench<half, int, raft::col_major>/11/manual_time        3.91 ms         3.96 ms          179 512#1000000
TransposeBench<half, int, raft::col_major>/12/manual_time       0.076 ms        0.112 ms         9122 1024#10000
TransposeBench<half, int, raft::col_major>/13/manual_time       0.796 ms        0.833 ms          878 1024#100000
TransposeBench<half, int, raft::col_major>/14/manual_time        7.95 ms         8.03 ms           84 1024#1000000

```

Reply:
@rhdong lets create an issue for Artem’s suggested change and reference it in a todo comment in the corresponding kernel in the code. I think we should investigate this for sure so that we are utilizing math libs where at all possible (and not having to maintain both math libs and our own custom impls) but I do not think the further investigation should hold up this PR.

Reply:
> @rhdong lets create an issue for Artem’s suggested change and reference it in a todo comment in the corresponding kernel in the code. I think we should investigate this for sure so that we are utilizing math libs where at all possible (and not having to maintain both math libs and our own custom impls) but I do not think the further investigation should hold up this PR.

Yeah, here it is: https://github.com/rapidsai/raft/issues/2436